{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                                                                     1\n",
       "journal_id                                                             1\n",
       "name                   ElasticBLAST: accelerating sequence search via...\n",
       "pmid                                                                 NaN\n",
       "pmc                                                             10040096\n",
       "publisher_id                                                        5245\n",
       "doi                                           10.1186/s12859-023-05245-9\n",
       "subject                                                         Software\n",
       "published_date                                                2023-03-26\n",
       "received_date                                                 2023-01-04\n",
       "accepted_date                                                 2023-03-21\n",
       "license_type                                                        None\n",
       "copyright_statement    Â© This is a U.S. Government work and not under...\n",
       "keywords               BLAST;Cloud computing;Alignment;Kubernetes;AWS...\n",
       "repositories           https://github.com/ncbi/elastic-blast;https://...\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import sqlite3\n",
    "\n",
    "# Dataset of a Study of Computational reproducibility of Jupyter notebooks from biomedical publications\n",
    "# https://zenodo.org/records/8226725\n",
    "with sqlite3.connect(\n",
    "    Path(\n",
    "        \"~/Downloads/computational-reproducibility-pmc/computational-reproducibility-pmc/analyses/db.sqlite\"\n",
    "    ).expanduser()\n",
    ") as conn:\n",
    "    df = pd.read_sql(\"SELECT * FROM ARTICLE\", conn)\n",
    "\n",
    "df.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3437\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doi</th>\n",
       "      <th>repo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5386</th>\n",
       "      <td>10.3389/fnbot.2019.00065</td>\n",
       "      <td>https://github.com/poppy-project/pypot</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6600</th>\n",
       "      <td>10.1099/mgen.0.000071</td>\n",
       "      <td>https://github.com/dmnfarrell/gordon-group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>10.1016/j.csbj.2022.08.035</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             doi                                        repo\n",
       "5386    10.3389/fnbot.2019.00065      https://github.com/poppy-project/pypot\n",
       "6600       10.1099/mgen.0.000071  https://github.com/dmnfarrell/gordon-group\n",
       "1204  10.1016/j.csbj.2022.08.035                                            "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a pairing of the doi and the repo\n",
    "repo_doi_pairs = []\n",
    "\n",
    "df = df.dropna(subset=[\"doi\", \"repositories\"])\n",
    "print(len(df))\n",
    "\n",
    "for _, row in df.iterrows():\n",
    "    # Convert repos to list\n",
    "    repos = row[\"repositories\"].split(\";\")\n",
    "    \n",
    "    # Add each repo to the list\n",
    "    for repo in repos:\n",
    "        repo_doi_pairs.append({\n",
    "            \"doi\": row[\"doi\"],\n",
    "            \"repo\": repo,\n",
    "        })\n",
    "\n",
    "repo_doi_pairs = pd.DataFrame(repo_doi_pairs)\n",
    "repo_doi_pairs.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fce729f904ea4d07b6c20ec749849ac1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/50 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corpus_id</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>repo</th>\n",
       "      <th>tldr</th>\n",
       "      <th>primary_topic</th>\n",
       "      <th>secondary_topic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>212736837</td>\n",
       "      <td>doi:10.7717/peerj.10316</td>\n",
       "      <td>A Bayesian brain model of adaptive behavior: a...</td>\n",
       "      <td>Adaptive behavior emerges through a dynamic in...</td>\n",
       "      <td>https://github.com/stefanradev93/BayesFlow</td>\n",
       "      <td>This work proposes and validate a new computat...</td>\n",
       "      <td>Biology</td>\n",
       "      <td>Medicine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>237149049</td>\n",
       "      <td>doi:10.1128/mSystems.00811-21</td>\n",
       "      <td>Separation of Donor and Recipient Microbial Di...</td>\n",
       "      <td>We assumed that the enrichment of successful g...</td>\n",
       "      <td>https://github.com/ivartb/RECAST</td>\n",
       "      <td>The recipient intestine colonization analysis ...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Biology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>236504044</td>\n",
       "      <td>doi:10.3389/fpls.2021.715309</td>\n",
       "      <td>An Affordable Image-Analysis Platform to Accel...</td>\n",
       "      <td>Recent technical advances in the computer-visi...</td>\n",
       "      <td>https://github.com/tensorflow/models</td>\n",
       "      <td>A platform that allows real-time stomata detec...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Computer Science</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    corpus_id                            doi  \\\n",
       "44  212736837        doi:10.7717/peerj.10316   \n",
       "38  237149049  doi:10.1128/mSystems.00811-21   \n",
       "8   236504044   doi:10.3389/fpls.2021.715309   \n",
       "\n",
       "                                                title  \\\n",
       "44  A Bayesian brain model of adaptive behavior: a...   \n",
       "38  Separation of Donor and Recipient Microbial Di...   \n",
       "8   An Affordable Image-Analysis Platform to Accel...   \n",
       "\n",
       "                                             abstract  \\\n",
       "44  Adaptive behavior emerges through a dynamic in...   \n",
       "38  We assumed that the enrichment of successful g...   \n",
       "8   Recent technical advances in the computer-visi...   \n",
       "\n",
       "                                          repo  \\\n",
       "44  https://github.com/stefanradev93/BayesFlow   \n",
       "38            https://github.com/ivartb/RECAST   \n",
       "8         https://github.com/tensorflow/models   \n",
       "\n",
       "                                                 tldr primary_topic  \\\n",
       "44  This work proposes and validate a new computat...       Biology   \n",
       "38  The recipient intestine colonization analysis ...      Medicine   \n",
       "8   A platform that allows real-time stomata detec...      Medicine   \n",
       "\n",
       "     secondary_topic  \n",
       "44          Medicine  \n",
       "38           Biology  \n",
       "8   Computer Science  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_of_pairs = repo_doi_pairs.sample(50)\n",
    "\n",
    "from rs_graph.data.enrichment.semantic_scholar import get_extended_paper_details\n",
    "\n",
    "# Get extended paper details\n",
    "sample_of_pairs[\"doi\"] = sample_of_pairs.doi.apply(lambda x: f\"doi:{x}\")\n",
    "sample_of_pairs_details = get_extended_paper_details(\n",
    "    sample_of_pairs.doi.tolist(),\n",
    "    filter_out_nones=False,\n",
    ")\n",
    "\n",
    "# Attach paper details to truede as columns\n",
    "biosciences_notebook_repro = []\n",
    "for repo, paper_detail in zip(sample_of_pairs.repo, sample_of_pairs_details, strict=True):\n",
    "    if paper_detail is not None:\n",
    "        biosciences_notebook_repro.append(\n",
    "            {\n",
    "                \"corpus_id\": paper_detail.corpus_id,\n",
    "                \"doi\": paper_detail.doi,\n",
    "                \"title\": paper_detail.title,\n",
    "                \"abstract\": paper_detail.abstract,\n",
    "                \"repo\": repo,\n",
    "                \"tldr\": paper_detail.tldr,\n",
    "                \"primary_topic\": paper_detail.primary_topic,\n",
    "                \"secondary_topic\": paper_detail.secondary_topic,\n",
    "            }\n",
    "        )\n",
    "\n",
    "biosciences_notebook_repro = pd.DataFrame(biosciences_notebook_repro)\n",
    "biosciences_notebook_repro.sample(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Computer Science         16\n",
       "Biology                  14\n",
       "Medicine                 11\n",
       "Chemistry                 2\n",
       "Environmental Science     2\n",
       "Geography                 1\n",
       "Physics                   1\n",
       "Engineering               1\n",
       "Name: singular_topic, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a column for \"singular_topic\" which is preferred to be \"secondary_topic\"\n",
    "# but if that is None then use \"primary_topic\"\n",
    "biosciences_notebook_repro[\"singular_topic\"] = (\n",
    "    biosciences_notebook_repro.secondary_topic.fillna(\n",
    "        biosciences_notebook_repro.primary_topic,\n",
    "    )\n",
    ")\n",
    "\n",
    "# Drop papers without a singular topic\n",
    "biosciences_notebook_repro = biosciences_notebook_repro.dropna(subset=[\"singular_topic\"])\n",
    "biosciences_notebook_repro.singular_topic.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|ââââââââââ| 48/48 [00:53<00:00,  1.11s/it]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(40, 10)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each repo, get the reademe (if it exists) and the description.\n",
    "from dotenv import load_dotenv\n",
    "from ghapi.all import GhApi\n",
    "from rs_graph.data.enrichment.github import get_repo_parts_from_url\n",
    "import base64\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "api = GhApi()\n",
    "\n",
    "def get_repo_readme(repo: str) -> str | None:\n",
    "    repo_parts = get_repo_parts_from_url(repo)\n",
    "    try:\n",
    "        b64_encoded_readme = api.repos.get_readme(\n",
    "                owner=repo_parts.owner,\n",
    "                repo=repo_parts.repo,\n",
    "            )[\"content\"]\n",
    "        \n",
    "        return base64.b64decode(b64_encoded_readme).decode(\"utf-8\")\n",
    "    except:\n",
    "        return None\n",
    "    \n",
    "# Get the READMEs and then then embeddings for each README\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "\n",
    "readmes = []\n",
    "for repo in tqdm(biosciences_notebook_repro[\"repo\"], total=len(biosciences_notebook_repro)):\n",
    "    readmes.append(get_repo_readme(repo))\n",
    "    time.sleep(0.85)\n",
    "\n",
    "# Attach to papers\n",
    "biosciences_notebook_repro[\"readme\"] = readmes\n",
    "\n",
    "# Drop nulls\n",
    "biosciences_notebook_repro = biosciences_notebook_repro.dropna(subset=[\"readme\"])\n",
    "biosciences_notebook_repro.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "015005aa2400483cb05a705b9513e36c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Batches:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(40, 768)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the embeddings for each README\n",
    "from sentence_transformers import SentenceTransformer\n",
    "model = SentenceTransformer(\"all-mpnet-base-v2\")\n",
    "\n",
    "readme_embeddings = model.encode(biosciences_notebook_repro[\"readme\"].tolist(), show_progress_bar=True)\n",
    "readme_embeddings.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corpus_id</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>repo</th>\n",
       "      <th>tldr</th>\n",
       "      <th>primary_topic</th>\n",
       "      <th>secondary_topic</th>\n",
       "      <th>singular_topic</th>\n",
       "      <th>readme</th>\n",
       "      <th>readme_embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>46862333</td>\n",
       "      <td>doi:10.1016/j.cels.2017.11.014</td>\n",
       "      <td>Enhancing Evolutionary Couplings with Deep Con...</td>\n",
       "      <td>None</td>\n",
       "      <td>https://github.com/soedinglab/CCMpred</td>\n",
       "      <td>DeepContact is introduced, a convolutional neu...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td># CCMpred\\n\\n[![Travis](https://img.shields.io...</td>\n",
       "      <td>[-0.00813202653080225, -0.02442438155412674, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>231149379</td>\n",
       "      <td>doi:10.3389/fbioe.2020.612832</td>\n",
       "      <td>Multi-Omics Driven Metabolic Network Reconstru...</td>\n",
       "      <td>An oleaginous yeast Rhodosporidium toruloides ...</td>\n",
       "      <td>https://github.com/stcoradetti/RBseq</td>\n",
       "      <td>An oleaginous yeast Rhodosporidium toruloides ...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Environmental Science</td>\n",
       "      <td>Environmental Science</td>\n",
       "      <td># RBseq\\nTrack fitness of deletion mutants wit...</td>\n",
       "      <td>[0.0118949543684721, 0.007754592224955559, 0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>235718678</td>\n",
       "      <td>doi:10.1186/s13321-021-00525-z</td>\n",
       "      <td>Nonadditivity in public and inhouse data: impl...</td>\n",
       "      <td>None</td>\n",
       "      <td>https://github.com/KramerChristian/Nonadditivi...</td>\n",
       "      <td>Nonadditivity (NA) analysis should be applied ...</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Medicine</td>\n",
       "      <td># Nonadditivity analysis\\n\\n[![Code style: bla...</td>\n",
       "      <td>[0.02291882038116455, 0.0009123127674683928, -...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    corpus_id                             doi  \\\n",
       "19   46862333  doi:10.1016/j.cels.2017.11.014   \n",
       "33  231149379   doi:10.3389/fbioe.2020.612832   \n",
       "6   235718678  doi:10.1186/s13321-021-00525-z   \n",
       "\n",
       "                                                title  \\\n",
       "19  Enhancing Evolutionary Couplings with Deep Con...   \n",
       "33  Multi-Omics Driven Metabolic Network Reconstru...   \n",
       "6   Nonadditivity in public and inhouse data: impl...   \n",
       "\n",
       "                                             abstract  \\\n",
       "19                                               None   \n",
       "33  An oleaginous yeast Rhodosporidium toruloides ...   \n",
       "6                                                None   \n",
       "\n",
       "                                                 repo  \\\n",
       "19              https://github.com/soedinglab/CCMpred   \n",
       "33               https://github.com/stcoradetti/RBseq   \n",
       "6   https://github.com/KramerChristian/Nonadditivi...   \n",
       "\n",
       "                                                 tldr     primary_topic  \\\n",
       "19  DeepContact is introduced, a convolutional neu...          Medicine   \n",
       "33  An oleaginous yeast Rhodosporidium toruloides ...          Medicine   \n",
       "6   Nonadditivity (NA) analysis should be applied ...  Computer Science   \n",
       "\n",
       "          secondary_topic         singular_topic  \\\n",
       "19       Computer Science       Computer Science   \n",
       "33  Environmental Science  Environmental Science   \n",
       "6                Medicine               Medicine   \n",
       "\n",
       "                                               readme  \\\n",
       "19  # CCMpred\\n\\n[![Travis](https://img.shields.io...   \n",
       "33  # RBseq\\nTrack fitness of deletion mutants wit...   \n",
       "6   # Nonadditivity analysis\\n\\n[![Code style: bla...   \n",
       "\n",
       "                                     readme_embedding  \n",
       "19  [-0.00813202653080225, -0.02442438155412674, -...  \n",
       "33  [0.0118949543684721, 0.007754592224955559, 0.0...  \n",
       "6   [0.02291882038116455, 0.0009123127674683928, -...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "# Attach back to dataframe\n",
    "biosciences_notebook_repro[\"readme_embedding\"] = [np.array(vec) for vec in readme_embeddings.tolist()]\n",
    "biosciences_notebook_repro.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8029193ed9934219906cd9aa7da65002",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Batches:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corpus_id</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>repo</th>\n",
       "      <th>tldr</th>\n",
       "      <th>primary_topic</th>\n",
       "      <th>secondary_topic</th>\n",
       "      <th>singular_topic</th>\n",
       "      <th>readme</th>\n",
       "      <th>readme_embedding</th>\n",
       "      <th>paper_embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>245916661</td>\n",
       "      <td>doi:10.1016/j.isci.2022.104769</td>\n",
       "      <td>Practical and thermodynamic constraints on ele...</td>\n",
       "      <td>None</td>\n",
       "      <td>https://github.com/barstowlab/electroco2</td>\n",
       "      <td>It is demonstrated that electromicrobial produ...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Biology</td>\n",
       "      <td>Biology</td>\n",
       "      <td># electroCO2\\n\\nBuz Barstow\\n2021-12-26\\n\\nCod...</td>\n",
       "      <td>[0.04969074949622154, -0.002880058716982603, -...</td>\n",
       "      <td>[0.05940726771950722, 0.021711746230721474, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>235718678</td>\n",
       "      <td>doi:10.1186/s13321-021-00525-z</td>\n",
       "      <td>Nonadditivity in public and inhouse data: impl...</td>\n",
       "      <td>None</td>\n",
       "      <td>https://github.com/KramerChristian/Nonadditivi...</td>\n",
       "      <td>Nonadditivity (NA) analysis should be applied ...</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Medicine</td>\n",
       "      <td># Nonadditivity analysis\\n\\n[![Code style: bla...</td>\n",
       "      <td>[0.02291882038116455, 0.0009123127674683928, -...</td>\n",
       "      <td>[-0.04822631925344467, 0.057587385177612305, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>248442073</td>\n",
       "      <td>doi:10.3390/life12050648</td>\n",
       "      <td>The R Language: An Engine for Bioinformatics a...</td>\n",
       "      <td>The R programming language is approaching its ...</td>\n",
       "      <td>https://github.com/jhallen/joes-sandbox</td>\n",
       "      <td>An historical chronicle of how R became what i...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>## Joe Allen's Open Source Projects\\n\\n### Pro...</td>\n",
       "      <td>[0.018987294286489487, 0.018373312428593636, -...</td>\n",
       "      <td>[-0.029452510178089142, -0.0029465905390679836...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    corpus_id                             doi  \\\n",
       "21  245916661  doi:10.1016/j.isci.2022.104769   \n",
       "6   235718678  doi:10.1186/s13321-021-00525-z   \n",
       "9   248442073        doi:10.3390/life12050648   \n",
       "\n",
       "                                                title  \\\n",
       "21  Practical and thermodynamic constraints on ele...   \n",
       "6   Nonadditivity in public and inhouse data: impl...   \n",
       "9   The R Language: An Engine for Bioinformatics a...   \n",
       "\n",
       "                                             abstract  \\\n",
       "21                                               None   \n",
       "6                                                None   \n",
       "9   The R programming language is approaching its ...   \n",
       "\n",
       "                                                 repo  \\\n",
       "21           https://github.com/barstowlab/electroco2   \n",
       "6   https://github.com/KramerChristian/Nonadditivi...   \n",
       "9             https://github.com/jhallen/joes-sandbox   \n",
       "\n",
       "                                                 tldr     primary_topic  \\\n",
       "21  It is demonstrated that electromicrobial produ...          Medicine   \n",
       "6   Nonadditivity (NA) analysis should be applied ...  Computer Science   \n",
       "9   An historical chronicle of how R became what i...          Medicine   \n",
       "\n",
       "     secondary_topic    singular_topic  \\\n",
       "21           Biology           Biology   \n",
       "6           Medicine          Medicine   \n",
       "9   Computer Science  Computer Science   \n",
       "\n",
       "                                               readme  \\\n",
       "21  # electroCO2\\n\\nBuz Barstow\\n2021-12-26\\n\\nCod...   \n",
       "6   # Nonadditivity analysis\\n\\n[![Code style: bla...   \n",
       "9   ## Joe Allen's Open Source Projects\\n\\n### Pro...   \n",
       "\n",
       "                                     readme_embedding  \\\n",
       "21  [0.04969074949622154, -0.002880058716982603, -...   \n",
       "6   [0.02291882038116455, 0.0009123127674683928, -...   \n",
       "9   [0.018987294286489487, 0.018373312428593636, -...   \n",
       "\n",
       "                                      paper_embedding  \n",
       "21  [0.05940726771950722, 0.021711746230721474, -0...  \n",
       "6   [-0.04822631925344467, 0.057587385177612305, -...  \n",
       "9   [-0.029452510178089142, -0.0029465905390679836...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the embeddings for each paper abstract\n",
    "# or if the abstract is None, use the title\n",
    "paper_texts = (\n",
    "    biosciences_notebook_repro[\"abstract\"].fillna(\n",
    "        biosciences_notebook_repro[\"tldr\"]\n",
    "    ).fillna(biosciences_notebook_repro[\"title\"]).tolist()\n",
    ")\n",
    "\n",
    "# Get the embeddings for each paper abstract\n",
    "paper_embeddings = model.encode(paper_texts, show_progress_bar=True)\n",
    "\n",
    "# Attach back to dataframe\n",
    "biosciences_notebook_repro[\"paper_embedding\"] = [np.array(vec) for vec in paper_embeddings.tolist()]\n",
    "biosciences_notebook_repro.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating interaction embeddings\n",
      "Predicting\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corpus_id</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>repo</th>\n",
       "      <th>tldr</th>\n",
       "      <th>primary_topic</th>\n",
       "      <th>secondary_topic</th>\n",
       "      <th>singular_topic</th>\n",
       "      <th>readme</th>\n",
       "      <th>readme_embedding</th>\n",
       "      <th>paper_embedding</th>\n",
       "      <th>interaction_embedding</th>\n",
       "      <th>prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>220305302</td>\n",
       "      <td>doi:10.1016/j.devcel.2022.01.008</td>\n",
       "      <td>A single cell Arabidopsis root atlas reveals d...</td>\n",
       "      <td>None</td>\n",
       "      <td>https://github.com/ohlerlab/COPILOT</td>\n",
       "      <td>The results suggest the existence of an altern...</td>\n",
       "      <td>Biology</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Medicine</td>\n",
       "      <td># COPILOT (Cell preprOcessing PIpeline kaLlist...</td>\n",
       "      <td>[-0.06168226897716522, -0.04459858685731888, -...</td>\n",
       "      <td>[0.011394570581614971, -0.04304368793964386, 0...</td>\n",
       "      <td>[-0.0007028429674944686, 0.0019196876552355358...</td>\n",
       "      <td>match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>235308894</td>\n",
       "      <td>doi:10.1038/s42003-021-02153-7</td>\n",
       "      <td>Transcription factor enrichment analysis (TFEA...</td>\n",
       "      <td>None</td>\n",
       "      <td>https://github.com/Dowell-Lab/mumerge</td>\n",
       "      <td>Transcription factor enrichment analysis (TFEA...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>ï»¿# muMerge\\r\\n\\r\\nmuMerge is a tool for combin...</td>\n",
       "      <td>[-0.03200733661651611, -0.020369168370962143, ...</td>\n",
       "      <td>[-0.014159103855490685, -0.0010391349205747247...</td>\n",
       "      <td>[0.00045319520329090146, 2.116631415733294e-05...</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>208567113</td>\n",
       "      <td>doi:10.1093/bioinformatics/btaa624</td>\n",
       "      <td>Style transfer with variational autoencoders i...</td>\n",
       "      <td>The transcriptomic data is being frequently us...</td>\n",
       "      <td>https://github.com/jdrudolph/goenrich</td>\n",
       "      <td>The proposed solution is based on Variational ...</td>\n",
       "      <td>Medicine</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>Computer Science</td>\n",
       "      <td>goenrich\\n========\\n\\n.. image:: https://badge...</td>\n",
       "      <td>[-0.03388982266187668, 0.03242969140410423, -0...</td>\n",
       "      <td>[-0.03258673474192619, 0.132096529006958, -0.0...</td>\n",
       "      <td>[0.0011043586615334944, 0.0042838496712489516,...</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    corpus_id                                 doi  \\\n",
       "45  220305302    doi:10.1016/j.devcel.2022.01.008   \n",
       "30  235308894      doi:10.1038/s42003-021-02153-7   \n",
       "26  208567113  doi:10.1093/bioinformatics/btaa624   \n",
       "\n",
       "                                                title  \\\n",
       "45  A single cell Arabidopsis root atlas reveals d...   \n",
       "30  Transcription factor enrichment analysis (TFEA...   \n",
       "26  Style transfer with variational autoencoders i...   \n",
       "\n",
       "                                             abstract  \\\n",
       "45                                               None   \n",
       "30                                               None   \n",
       "26  The transcriptomic data is being frequently us...   \n",
       "\n",
       "                                     repo  \\\n",
       "45    https://github.com/ohlerlab/COPILOT   \n",
       "30  https://github.com/Dowell-Lab/mumerge   \n",
       "26  https://github.com/jdrudolph/goenrich   \n",
       "\n",
       "                                                 tldr primary_topic  \\\n",
       "45  The results suggest the existence of an altern...       Biology   \n",
       "30  Transcription factor enrichment analysis (TFEA...      Medicine   \n",
       "26  The proposed solution is based on Variational ...      Medicine   \n",
       "\n",
       "     secondary_topic    singular_topic  \\\n",
       "45          Medicine          Medicine   \n",
       "30  Computer Science  Computer Science   \n",
       "26  Computer Science  Computer Science   \n",
       "\n",
       "                                               readme  \\\n",
       "45  # COPILOT (Cell preprOcessing PIpeline kaLlist...   \n",
       "30  ï»¿# muMerge\\r\\n\\r\\nmuMerge is a tool for combin...   \n",
       "26  goenrich\\n========\\n\\n.. image:: https://badge...   \n",
       "\n",
       "                                     readme_embedding  \\\n",
       "45  [-0.06168226897716522, -0.04459858685731888, -...   \n",
       "30  [-0.03200733661651611, -0.020369168370962143, ...   \n",
       "26  [-0.03388982266187668, 0.03242969140410423, -0...   \n",
       "\n",
       "                                      paper_embedding  \\\n",
       "45  [0.011394570581614971, -0.04304368793964386, 0...   \n",
       "30  [-0.014159103855490685, -0.0010391349205747247...   \n",
       "26  [-0.03258673474192619, 0.132096529006958, -0.0...   \n",
       "\n",
       "                                interaction_embedding prediction  \n",
       "45  [-0.0007028429674944686, 0.0019196876552355358...      match  \n",
       "30  [0.00045319520329090146, 2.116631415733294e-05...   no_match  \n",
       "26  [0.0011043586615334944, 0.0042838496712489516,...   no_match  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from skops import io as skio\n",
    "\n",
    "# Load the model\n",
    "model = skio.load(\"repo-paper-matching-model.skops\", trusted=True)\n",
    "\n",
    "# Create interaction embeddings and train logistic regression CV\n",
    "print(\"Creating interaction embeddings\")\n",
    "biosciences_notebook_repro[\"interaction_embedding\"] = biosciences_notebook_repro.apply(\n",
    "    lambda row: row.paper_embedding * row.readme_embedding,\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "print(\"Predicting\")\n",
    "biosciences_notebook_repro[\"prediction\"] = model.predict(\n",
    "    biosciences_notebook_repro[\"interaction_embedding\"].tolist()\n",
    ")\n",
    "\n",
    "biosciences_notebook_repro.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://github.com/zxing/zxing\n",
      "Performant barcode decoding for herbarium specimen images using vectorâassisted region proposals (VARP)\n",
      "doi:10.1002/aps3.11436\n",
      "\n",
      "https://github.com/stcoradetti/RBseq\n",
      "Multi-Omics Driven Metabolic Network Reconstruction and Analysis of Lignocellulosic Carbon Utilization in Rhodosporidium toruloides\n",
      "doi:10.3389/fbioe.2020.612832\n",
      "\n",
      "https://github.com/soedinglab/CCMpred\n",
      "Enhancing Evolutionary Couplings with Deep Convolutional Neural Networks\n",
      "doi:10.1016/j.cels.2017.11.014\n",
      "\n",
      "https://github.com/r3fang/SnapATAC\n",
      "Single nucleus multi-omics identifies human cortical cell regulatory genome diversity\n",
      "doi:10.1016/j.xgen.2022.100107\n",
      "\n",
      "https://github.com/collaborativebioinformatics/cov2db\n",
      "The third international hackathon for applying insights into large-scale genomic composition to use cases in a wide range of organisms\n",
      "doi:10.12688/f1000research.110194.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Print dataframe of matches\n",
    "for _, row in biosciences_notebook_repro[biosciences_notebook_repro.prediction == \"match\"].sample(5).iterrows():\n",
    "    print(row.repo)\n",
    "    print(row.title)\n",
    "    print(row.doi)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://github.com/owlcs/owlapi\n",
      "Semantic similarity and machine learning with ontologies\n",
      "doi:10.1093/bib/bbaa199\n",
      "\n",
      "https://github.com/BlueBrain/BlueBrainGraph\n",
      "A Machine-Generated View of the Role of Blood Glucose Levels in the Severity of COVID-19\n",
      "doi:10.3389/fpubh.2021.695139\n",
      "\n",
      "https://github.com/information-artifact-ontology/IAO\n",
      "Community standards for open cell migration data\n",
      "doi:10.1093/gigascience/giaa041\n",
      "\n",
      "https://github.com/duderstadt-lab/Born-to-slide\n",
      "Mobile origin-licensing factors confer resistance to conflicts with RNA polymerase\n",
      "doi:10.1016/j.celrep.2022.110531\n",
      "\n",
      "https://github.com/Dowell-Lab/mumerge\n",
      "Transcription factor enrichment analysis (TFEA) quantifies the activity of multiple transcription factors from a single experiment\n",
      "doi:10.1038/s42003-021-02153-7\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Print dataframe of matches\n",
    "for _, row in biosciences_notebook_repro[biosciences_notebook_repro.prediction == \"no_match\"].sample(5).iterrows():\n",
    "    print(row.repo)\n",
    "    print(row.title)\n",
    "    print(row.doi)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
